include/deal.II/matrix_free/cuda_matrix_free.templates.h

```diff
@@ -1032,17 +1067,44 @@ namespace CUDAWrappers
     if (src.get_partitioner().get() == partitioner.get() &&
         dst.get_partitioner().get() == partitioner.get())
       {
-        src.update_ghost_values();
 
         // Execute the loop on the cells
-        for (unsigned int i = 0; i < n_colors; ++i)
+        if(use_ghost_coloring)
+        {
+          src.update_ghost_values_start(0);
           internal::apply_kernel_shmem<dim, Number, Functor>
-            <<<grid_dim[i], block_dim[i]>>>(func,
-                                            get_data(i),
+            <<<grid_dim[0], block_dim[0]>>>(func,
+                                            get_data(0),
                                             src.get_values(),
                                             dst.get_values());
-        dst.compress(VectorOperation::add);
-        src.zero_out_ghosts();
+          src.update_ghost_values_finish();
+          internal::apply_kernel_shmem<dim, Number, Functor>
```

Comment:
I think this kernel should be preceded and followed by `cudaDeviceSynchronize()` because of CUDA-aware MPI. How do you know that the data exchanged during the update_ghost/compress has been computed by the time the computation is done? If you don't use CUDA-aware MPI, we know that everything is using the same stream so it implicitly synchronize but we don't know when using CUDA-aware MPI. This also explains the problem you saw: you have a race condition and using more colors changes the time require to loop over the domain.

Reply:
@Rombur Thanks for the debugging! 

I am wondering which version of the code are you using? Are you using this version or the version on https://github.com/peterrum/dealii/tree/dealii-on-gpu? I would think 369a5b9ab5c203d651a1f435c3f01d152dcb3ef5 is working. However, what we observed at the workshop that the memcopy DtoH and HtoD was not overlapped, that is why we introduced the streams in 75988028d91e70095db24df61efebba395a2d664. I am afraid here we indeed need the synchronizations (before and after the second block - although I hoped that it is implicitly synchronized by the default stream).

Reply:
I found the bug. You can't see it if you use CUDA-aware MPI but funny enough you can see it in serial if you use regular MPI. In that case, we copy all the data to host, perform the communication, and then copy **all** the data back to the device. This means that if we change the data on the device, it is overwritten by the data from the host.

Reply:
Does this mean that the HtoD in `update_ghost_values_finish()` overrides values that have been computed in the first block? Do I understand it correctly? But how is that possible? Only the ghost values are copied around and that should be covered by the coloring (if it is correct), shouldn't it?

Reply:
No actually the problem is [here](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/la_parallel_vector.templates.h#L961-L965) in `compress_finish`. The values computed in block 2 are not taken into account.

Reply:
So the problem is that when you don't use CUDA-aware MPI, the function `import_from_ghosted_array_finish` uses the data that has been copied to the host during the call to `compress_start`. If you update the data on the device after the call to `compress_start`, the host is not updated and you lose the new values. I need to double check but I think the solution is to copy the ghost data to the device, then do the call to `import_from_ghosted_array_finish`, and change `import_from_ghosted_array_finish` so that there is one CUDA path and one host path regardless of CUDA-aware MPI.

Reply:
This is the `if defined DEAL_II_COMPILER_CUDA_AWARE && !defined  DEAL_II_MPI_WITH_CUDA_SUPPORT` branch. In this configuration the whole vector is copied from and to the host. But that is not the case for `if defined DEAL_II_COMPILER_CUDA_AWARE && defined  DEAL_II_MPI_WITH_CUDA_SUPPORT`, isn't it?

Reply:
I think what I wanted to do [here](https://github.com/dealii/dealii/commit/369a5b9ab5c203d651a1f435c3f01d152dcb3ef5) was not write `#ifdef DEAL_II_COMPILER_CUDA_AWARE` but `#ifdef DEAL_II_MPI_WITH_CUDA_SUPPORT`. Sorry about this.

The whole purpose of the cude-aware MPI implementation is not to copy the whole vector.

Reply:
> This is the if defined DEAL_II_COMPILER_CUDA_AWARE && !defined DEAL_II_MPI_WITH_CUDA_SUPPORT branch. In this configuration the whole vector is copied from and to the host. But that is not the case for if defined DEAL_II_COMPILER_CUDA_AWARE && defined DEAL_II_MPI_WITH_CUDA_SUPPORT, isn't it?

That's right.

Reply:
> That's right.

I noticed your message to late...